lib/srv/ctx.go

```diff
@@ -676,24 +678,74 @@ func (c *ServerContext) String() string {
 	return fmt.Sprintf("ServerContext(%v->%v, user=%v, id=%v)", c.ServerConn.RemoteAddr(), c.ServerConn.LocalAddr(), c.ServerConn.User(), c.id)
 }
 
-// ExecCommand takes a *ServerContext and extracts the parts needed to create
-// an *execCommand which can be re-sent to Teleport.
-func (c *ServerContext) ExecCommand() (*ExecCommand, error) {
-	var pamEnabled bool
-	var pamServiceName string
-	var pamUseAuth bool
+func getPAMConfig(c *ServerContext) (*PAMConfig, error) {
+	// PAM should be disabled.
+	if c.srv.Component() != teleport.ComponentNode {
+		return nil, nil
+	}
+
+	localPAMConfig, err := c.srv.GetPAM()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-	// If this code is running on a node, check if PAM is enabled or not.
-	if c.srv.Component() == teleport.ComponentNode {
-		conf, err := c.srv.GetPAM()
+	// We use nil/empty to figure out if PAM is disabled later.
+	if !localPAMConfig.Enabled {
+		return nil, nil
+	}
+
+	// Fill in the environment variables from the config and interpolate them if needed.
+	environment := make(map[string]string)
+	if localPAMConfig.Environment != nil {
+		user, err := c.srv.GetAccessPoint().GetUser(c.Identity.TeleportUser, false)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
-		pamEnabled = conf.Enabled
-		pamServiceName = conf.ServiceName
-		pamUseAuth = conf.UsePAMAuth
+
+		traits := user.GetTraits()
+
+		for key, value := range localPAMConfig.Environment {
+			expr, err := parse.NewExpression(value)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			if expr.Namespace() != teleport.TraitExternalPrefix {
+				return nil, trace.BadParameter("PAM environment interpolation only support external traits, found %q", value)
+			}
+
+			result, err := expr.Interpolate(traits)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			environment[key] = strings.Join(result, " ")
+		}
 	}
 
+	return &PAMConfig{
+		UsePAMAuth:  localPAMConfig.UsePAMAuth,
+		ServiceName: localPAMConfig.ServiceName,
+		Environment: environment,
+	}, nil
+}
+
+func getPAMEnvironment(c *ExecCommand) *map[string]string {
```

Comment:
```suggestion
func getPAMEnvironment(c *ExecCommand) map[string]string {
```
as you can still return nil for empty since maps are reference types.